### PR TITLE
Return status code exceptions via CompletableResultCode in GrpcExporter and HttpExporter

### PR DESCRIPTION
**Pull request by jarrodrobins**
Building off of @jack-berg's PR #6348, this PR uses the newly added `CompletableResultCode`'s `failExceptionally` method to return status code errors from the `GrpcExporter` and `HttpExporter`.

The intention is to expose the status code upstream so it can be responded to, rather than simply have an 'unknown' failure. This was requested #6306. There was some discussion there around parsing response bodies which sounded like it was quite complicated; I have just focused on the basic case of returning the HTTP or GRPC status code.

This is my first PR into OpenTelemetry so please let me know if what I've done should be done differently or if you have suggestions for a better way. 

I initially had separate exceptions for the `GrpcExporter` and `HttpExporter`s, but this seemed like overkill considering they did the same thing. Additionally, I wasn't sure if there was a better package for the exception itself - I've put it in the `internal` package currently with the associated code.

Please take a look - happy to make changes as necesssary.

### CODE REVIEW

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/ExporterStatusException.java
Comment (jack-berg) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/ExporterStatusException.java:17]: I think we need to be able to have different representations for http and grpc errors, and for exports that fail with a response from the server vs. exports that fail exceptionally. I sketched out a exception class hierarchy. What do you think?

```
package io.opentelemetry.exporter.internal;

import io.opentelemetry.exporter.internal.grpc.GrpcResponse;
import io.opentelemetry.exporter.internal.http.HttpSender;
import javax.annotation.Nullable;

public abstract class FailedExportException extends Exception {

  private FailedExportException(@Nullable Throwable cause) {
    super(cause);
  }

  /** Indicates an HTTP export failed after receiving a response from the server. */
  public static HttpExportException httpFailedWithResponse(HttpSender.Response response) {
    return new HttpExportException(response, null);
  }

  /** Indicates an HTTP export failed exceptionally without receiving a response from the server. */
  public static HttpExportException httpFailedExceptionally(Throwable cause) {
    return new HttpExportException(null, cause);
  }

  /** Indicates an HTTP export failed after receiving a response from the server. */
  public static GrpcExportException grpcFailedWithResponse(GrpcResponse response) {
    return new GrpcExportException(response, null);
  }

  /** Indicates an HTTP export failed exceptionally without receiving a response from the server. */
  public static GrpcExportException grpcFailedExceptionally(Throwable cause) {
    return new GrpcExportException(null, cause);
  }

  /** Returns true if the export failed with a response from the server. */
  abstract boolean failedWithResponse();

  /** Represents the failure of an HTTP exporter. */
  public static final class HttpExportException extends FailedExportException {

    @Nullable private final HttpSender.Response response;
    @Nullable private final Throwable cause;

    private HttpExportException(@Nullable HttpSender.Response response, @Nullable Throwable cause) {
      super(cause);
      this.response = response;
      this.cause = cause;
    }

    @Override
    boolean failedWithResponse() {
      return response != null;
    }

    /**
     * Returns the response if the export failed with a response from the server, or null if the
     * export failed exceptionally with no response.
     */
    @Nullable
    public HttpSender.Response getResponse() {
      return response;
    }

    /**
     * Returns the exceptional cause of failure, or null if the export failed with a response from
     * the server.
     */
    @Nullable
    public Throwable getCause() {
      return cause;
    }
  }

  /** Represents the failure of a gRPC exporter. */
  public static final class GrpcExportException extends FailedExportException {

    @Nullable private final GrpcResponse response;
    @Nullable private final Throwable cause;

    private GrpcExportException(@Nullable GrpcResponse response, @Nullable Throwable cause) {
      super(cause);
      this.response = response;
      this.cause = cause;
    }

    @Override
    boolean failedWithResponse() {
      return response != null;
    }

    /**
     * Returns the response if the export failed with a response from the server, or null if the
     * export failed exceptionally with no response.
     */
    @Nullable
    public GrpcResponse getResponse() {
      return response;
    }

    /**
     * Returns the exceptional cause of failure, or null if the export failed with a response from
     * the server.
     */
    @Nullable
    public Throwable getCause() {
      return cause;
    }
  }
}
```

Comment (jarrodrobins) [reply exporters/common/src/main/java/io/opentelemetry/exporter/internal/ExporterStatusException.java]: This is great actually, and gives a lot more flexibility to clients.

**Comment by jarrodrobins** [comment]
Thanks @jack-berg, I've updated my PR to include your suggestion. I also broke out the logging into a separate method. Take a look, I think I'm handling the gRPC error codes correctly (I looked at what `OkHttpGrpcSender` was doing). Let me know if you need me to make any other changes.

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/FailedExportException.java
Comment (jack-berg) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/FailedExportException.java:36]:
```suggestion
  /** Indicates an gRPC export failed after receiving a response from the server. */
```
copy paste error on my part

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/FailedExportException.java
Comment (jack-berg) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/FailedExportException.java:41]:
```suggestion
  /** Indicates an gRPC export failed exceptionally without receiving a response from the server. */
```

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java
Comment (jack-berg) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java:79]: The semantics behind when we call `grpcFailedExceptionally` and `grpcFailedWithResponse` are pretty loose right now - why is unknown status `grpcFailedExceptionally` while the other status codes are `grpcFailedWithResponse`? Makes you think that `FailedExporterException.GrpcExportException` can have non-null responses for both `getResponse()` and `getCause()`. But I don't think this is actually the case. I do think its one or the other.

 I think the fault lies with the GrpcSender interface rather than anything in this PR. Note that this same problem doesn't exist for HttpExporter? I think the HttpSender interface got it right. Opened #6658 to refactor the GrpcSender interface to match HttpExporter, which would clean up the semantics here.

No need to have this PR be blocked on merging that. I can solve merge conflicts and clean up #6658 after merging this.

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java
Comment (jack-berg) [inline exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java:582]: Can we assert on the response code of the exception as well? Can we include this additional assertion in one of the existing tests like `error()` instead of introducing a new one?

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractHttpTelemetryExporterTest.java
Comment (jack-berg) [inline exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractHttpTelemetryExporterTest.java:548]: Can we assert that `getCause()` is non-null and `getResponse()` is null?

**Review by jack-berg — COMMENTED** [review]
Just a couple of small comments but this looks pretty good. Thanks!

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/FailedExportException.java
Comment (jarrodrobins) [reply exporters/common/src/main/java/io/opentelemetry/exporter/internal/FailedExportException.java:36]: Oops - easy fix

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java
Comment (jarrodrobins) [reply exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporter.java:79]: Yeah, I did feel this was a little odd but didn't want to delve too deep in this PR. Thanks for taking a look at fixing that.

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java
Comment (jarrodrobins) [reply exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java:582]: Yep no problem.

**Comment by jarrodrobins** [comment]
Thanks @jack-berg, I've made those changes.

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/http/HttpExporter.java
Comment (jack-berg) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/http/HttpExporter.java:78]: The fact that this `responseBody()` throws when called multiple times is a bug. There's nothing in the contract of `HttpSender.Response.responseBody()` that suggests that you can't call it multiple times. While this avoids the issue, I think we need to fix the root of the problem in `OkHttpHttpSender` with something like:
```
onResponse.accept(
    new Response() {
      @Nullable
      private byte[] bodyBytes;
      @Override
      public int statusCode() {
        return response.code();
      }

      @Override
      public String statusMessage() {
        return response.message();
      }

      @Override
      public byte[] responseBody() throws IOException {
        if (bodyBytes == null) {
          bodyBytes = body.bytes();
        }
        return bodyBytes;
      }
    });
```

Comment (jarrodrobins) [reply exporters/common/src/main/java/io/opentelemetry/exporter/internal/http/HttpExporter.java]: Alright, I've reverted that last change and have added this.

**Review by jack-berg — APPROVED** [review]

**Pull request merged**
